sycl/source/detail/scheduler/commands.cpp

```diff
@@ -2397,13 +2400,38 @@ static ur_result_t SetKernelParamsAndLaunch(
                               : Empty);
   }
 
-  auto setFunc = [&Adapter, Kernel, &DeviceImageImpl, &getMemAllocationFunc,
-                  &Queue](detail::ArgDesc &Arg, size_t NextTrueIndex) {
-    SetArgBasedOnType(Adapter, Kernel, DeviceImageImpl, getMemAllocationFunc,
-                      Queue->getContextImplPtr(), Arg, NextTrueIndex);
-  };
-
-  applyFuncOnFilteredArgs(EliminatedArgMask, Args, setFunc);
+  if (KernelFuncPtr && !KernelHasSpecialCaptures) {
+    auto setFunc = [&Adapter, Kernel,
+                    KernelFuncPtr](const detail::kernel_param_desc_t &ParamDesc,
+                                   size_t NextTrueIndex) {
+      const void *ArgPtr = (const char *)KernelFuncPtr + ParamDesc.offset;
+      switch (ParamDesc.kind) {
+      case kernel_param_kind_t::kind_std_layout: {
+        int Size = ParamDesc.info;
+        Adapter->call<UrApiKind::urKernelSetArgValue>(Kernel, NextTrueIndex,
+                                                      Size, nullptr, ArgPtr);
+        break;
+      }
+      case kernel_param_kind_t::kind_pointer: {
+        const void *Ptr = *static_cast<const void *const *>(ArgPtr);
+        Adapter->call<UrApiKind::urKernelSetArgPointer>(Kernel, NextTrueIndex,
+                                                        nullptr, Ptr);
+        break;
+      }
+      default:
+        throw std::runtime_error("Direct kernel argument copy failed.");
+      }
+    };
```

Comment:
Could it make sense to add a `SetArgBasedOnType` overload or variant that the old one can call if the arguments don't fall under the special types? Avoids a little code replication.

Reply:
Ah, I see what you mean. Yeah, I am not sure it's worth trying to repack the arguments.

Reply:
I think that could be possible if we had more special-casing in the submission paths, but I didn't want to embark on a redesign that big.

I might have misunderstood the current code, but I convinced myself that the `MArgs` vector as it's currently used could be populated either by extracting arguments from a lambda, or from the user calling `set_arg`/`set_args` to provide the arguments for a `sycl::kernel`.  If we pushed `MArgs` creation all the way down, I'm not sure how to handle the `set_arg`/`set_args` case.

I wrote this primarily to demonstrate that extracting things directly from the kernel function object was 1) possible; and 2) faster.  I'd love to make that the default path for everything, but I don't understand the runtime well enough!

Reply:
Can you describe how exactly you're testing performance of this? Either PR's description or maybe a comment near the fast path code. That would help whoever will try to generalize your approach for all submission paths to ensure your "fast path" doesn't regress.

Reply:
Good idea.  I've added a paragraph to the end of the PR description.